### PR TITLE
feat: wrap non-hoistable CommonJS modules into module concatenation

### PR DESCRIPTION
**Pull request by alexander-akait**
**Summary**

Follow-up to #21417. "Nice" CommonJS modules (a static list of `exports.x =` assignments) are already hoisted into module concatenation; "weird" ones — those that reassign `module.exports`, use `this` as the exports object, define exports via `Object.defineProperty`, or let the exports object escape — still bailed out and stayed behind the `__webpack_require__` runtime. This wraps such modules into the concatenation instead: each is rendered inside an IIFE with real `module`/`exports` objects, and its live `.exports` is exposed as the module's namespace so consumers bind to it directly.

Gated behind the existing `optimization.concatenateModules.commonjs` opt-in (on by default when concatenation runs). Modules that `require()` other modules, use `module.id`/`module.loaded`, or are sloppy-mode still bail out (require-target inlining is left for a follow-up). Verified behavior-preserving: every wrapped pattern produces output byte-behavior-identical to the non-concatenated baseline and matches Node's own CommonJS semantics.

Note: `types.d.ts` was updated by hand (a two-line `ConcatenatedModuleInfo` addition) because `yarn fix:special` could not run in my environment — its `webpack/tooling` dependency was unreachable; please re-run `yarn fix:special` to confirm the exact generated form.

Refs #17121

**What kind of change does this PR introduce?**

feat

**Did you add tests for your changes?**

Yes. `test/configCases/concatenate-modules/commonjs-wrapped/` covers 25 real-world export shapes (object/function/class/primitive/array reassignment, `this`, `Object.defineProperty`/`defineProperties`, getter objects, escaped exports, the `__esModule` flag, `module.exports = exports = {}` alias forms, local `exports` reassignment, multi/mixed reassignment, conditional/frozen exports, `this`-methods, live bindings). `commonjs-bailouts/` is extended with a whole `module.exports = require(...)` re-export and a circular CommonJS pair, asserting they still bail and run correctly.

**Does this PR introduce a breaking change?**

No. It only widens which modules concatenate; wrapped output is behavior-equivalent to today's, and the path is behind the existing opt-in.

**If relevant, what needs to be documented once your changes are merged or what have you already documented?**

n/a — `optimization.concatenateModules.commonjs` already documents CommonJS concatenation; this extends its coverage without new options.

**Use of AI**

AI (Claude) was used to help design the wrapping approach, write the implementation and tests, and adversarially verify every CommonJS export pattern against Node's semantics and the non-concatenated baseline. All changes were reviewed and validated against the full config test suite.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01CvYCNo2Ff4VxCwwHJSjw25)_

### CODE REVIEW

File: test/configCases/concatenate-modules/commonjs-wrapped/exports-local-reassign.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+"use strict";
+
+// reassigning the local `exports` binding does NOT change module.exports
+exports.keep = 1;
+exports = { gone: 2 };
```
[the remark this answers is not in the document]
Comment (alexander-akait) [reply test/configCases/concatenate-modules/commonjs-wrapped/exports-local-reassign.js:5]: This is intentional. The fixture exists specifically to verify that wrapping preserves CommonJS semantics for a reassigned local `exports` binding: `exports = { gone: 2 }` is a no-op on `module.exports`, so the wrapped module must export only `keep` and not `gone` — which the test asserts. Applying the suggestion (making `gone` a real export) would remove the exact edge case under test, so keeping the "useless" assignment here is deliberate.

---
_Generated by [Claude Code](https://claude.ai/code)_
